_posts/2016-01-25-composer-mailchimp-subscriptions.md

```diff
@@ -0,0 +1,92 @@
+- Managed the MailChimp library dependency via Composer Manager.
+
+#### Custom vs. Conrtib?
+But first, why didn't we just use the MailChimp contributed module? Contributed modules are often a great option and offer many benefits, such as security, maintenance, and flexibility.
+
+But there is a cost to installing all those contributed modules. As [*The Definitive Guide to Drupal 7*](http://definitivedrupal.org/) explains "The more modules you install, the worse your web site will perform."
+
+With each installed module comes more code to load and execute, and more memory consumption. And in some cases, contributed modules add complexity and features that just aren't necessary for the required task.
+
+In our case, the decision to go with a custom solution was easy:
+
+- The MailChimp contributed module
+was bloated with features we didn't need.
+- We were already using Composer Manager on the project to manage other module dependencies.
+- The custom module we were building already included logic to determine when to subscribe users to mailing lists (don't worry, we made sure they opted in!)
+
+All we needed was a simple, lightweight method for subscribing a given user to a specific MailChimp mailing list.
+
+#### Implementation
+We were able to achieve this by adding the [MailChimp PHP library](https://packagist.org/packages/mailchimp/mailchimp) as a dependency of our custom module. We were then able to make a simple call using the API to subscribe a user to the mailing list. We implemented this via the following code.
+
+First, in our module's root directory we created a `composer.json` file that specified the MailChimp PHP library as a dependency:
+
+{% highlight php %}
+{
+  "require": {
+    "mailchimp/mailchimp": "*"
+  }
+}
+{% endhighlight %}
+
+We then installed the Mailchimp API using the Composer Manager drush commands:
+
+{% highlight bash %}
+drush composer-json-rebuild
+drush composer-manager install
+{% endhighlight %}
+
+As explained in my [last post](/2015/10/15/composing-with-composer-manager.html), the first command builds  (or rebuilds) the consolidated project wide `composer.json` file and the second command installs the dependencies.
+
+Next, we created a function in our custom module to subscribe a user to a MailChimp mailing list.
+
+{% highlight php %}
+/**
+ * Add an email to a MailChimp list.
+ *
+ * @param string $api_key
+ *   The MailChimp API key.
+ * @param string $list_id
+ *   The MailChimp list id that the user should be subscribed to.
+ * @param string $email
+ *   The email address for the user being subscribed to the mailing list.
+ */
+function module_subscribe_user($api_key, $list_id, $email) {
+
+  $mailchimp = new Mailchimp($api_key);
+
+  try {
+    $result = $mailchimp->lists->subscribe($list_id, array('email' => $email));
+  }
+  catch(Exception $e) {
+    watchdog(WATCHDOG_WARNING, 'User not subscribed to ' . $e->getMessage());
```

Comment:
You know the code better wouldn't it be $list_id instead of $e->getMessage() ? e.g.